includes/class-content-toc.php

```diff
@@ -96,69 +99,112 @@ public function get_default_headers() {
+		 * Add `the_content` filter at 12, because `do_shortcode` is run at 11
+		 * and it means that we don't have any access to the posts content within
+		 * the shortcode callback, i.e. this function
+		 *
+		 * So:
+		 * 1) Shortcode is replaced with placeholder HTML
+		 * 2) `the_content` filter is added at priority 12, so that we can access post content
+		 *    after the shortcode has been processed/replaced
+		 * 3) The added filter is self removed, so it only runs on content that contains the shortcode
+		 */
+		add_filter( 'the_content', $func = function( $post_content ) use ( $shortcode_atts, &$func ) {
+
+			// Self remove just added filter, so it only runs whenever there is specified shortcode
+			remove_filter( 'the_content', $func, 12 );
+
+			// Process post content - insert TOC and anchors before headers
+			return TOC::get_instance()->filter_content( $post_content, $shortcode_atts );
+
+		}, 12 );
+
+		/**
+		 * Shortcode is substituted for HTML placeholder.
+		 * This is done, so that we can call `the_content` filter within
+		 * this shortcode function and self remove that filter.
+		 *
+		 * Currently this approach is buggy in WP core ref: https://core.trac.wordpress.org/ticket/17817
+		 * hence it requires a workaround.
+		 */
+		return $this->placeholder;
+	}
+
+	/**
+	 * Create string ' class=""' with specified class and suffix,
+	 * escaped and ready to use in HTML
+	 *
+	 * @param string $class  The class string
+	 * @param string $suffix Suffix, additional class string
+	 *
+	 * @return string        String ' class="{$class}-{$suffix}"' escaped and ready to use in HTML
+	 *                       If $class is empty/not specified, return original value
+	 */
+	protected function tag_class( $class, $suffix = '' ) {
+
+		if ( $class ) {
+
+			// Append suffix
+			if ( $suffix ) {
+				$class .= '-' . $suffix;
+			}
 
-			$title_html = sprintf( '<%1$s%2$s>%3$s</%1$s>',
-				esc_attr( $this->settings['title_tag'] ),
-				$this->tag_class( $this->settings['title_class'] ),
-				esc_html( $atts['title'] )
-				);
+			$class = sprintf( ' class="%s"', esc_attr( $class ) );
 		}
 
-		// TOC items HTML
-		$items_html = '';
-		foreach ( $items as $key => $item_match_arr ) {
+		return $class;
+	}
 
-			// Counter of items, starting at 1
-			$key_current = $key + 1;
+	/**
+	 * Callback for applying filter to post content
+	 *
+	 * Replaces content TOC placeholder with content TOC HTML and inserts anchor tags at headings
+	 *
+	 * @param string $post_content   The content HTML string, comming from `the_content` filter
+	 * @param array  $shortcode_atts The shortcode attributes, coming from `hm_content_toc` shortcode
+	 *
+	 * @return string
+	 */
+	public function filter_content( $post_content, $shortcode_atts ) {
 
-			// Stripped item text
-			$item_text = strip_tags( $item_match_arr[1] );
+		// Reset the counter
+		// TODO - Theo, could you explain a little bit more why? Is it due to it being singleton class?
+		$this->id_counter = 0;
```

Comment:
Allows for compatibility with archive pages. Counter is being incremented on the_content hook 11, by the_content hook 12, the previous value of the counter is no longer relevant because we have already parsed all the shortcodes for the current post. If we have 2 posts loading on a page load, then without resetting the counter here, the 2nd post will not have it's shortcode parsed. see https://github.com/dashaluna/hm-content-toc/pull/35/files#diff-a20892fca22525983b4f0537d0ea0ebaR115